HeliumVotePlugin/components/LockTokensModal.tsx

```diff
@@ -0,0 +1,362 @@
+  none = 'none',
+  cliff = 'cliff',
+  constant = 'constant',
+}
+
+const lockupInfosByType = {
+  [LockupKind.cliff]: [
+    'Tokens are locked for a fixed duration and are released in full at the end of it.',
+    'Vote weight declines linearly until release.',
+    'Example: You lock 10.000 tokens for two years. They are then unavailable for the next two years. After this time, you can withdraw them again.',
+  ],
+  [LockupKind.constant]: [
+    'Tokens are locked indefinitely. At any time you can start the unlock process which lasts for the initially chosen lockup duration.',
+    'Vote weight stays constant until you start the unlock process, then it declines linearly until release.',
+    'Example: You lock 10.000 tokens with a lockup duration of one year. After two years you decide to start the unlocking process. Another year after that, you can withdraw the tokens.',
+  ],
+}
+
+export interface LockTokensModalFormValues {
+  lockupKind: { value: LockupKind; display: string }
+  amount: number
+  lockupPeriod: { value: number; display: string }
+  lockupPeriodInDays: number
+}
+
+export const LockTokensModal: React.FC<{
+  isOpen: boolean
+  mode?: 'lock' | 'extend'
+  minLockupTimeInDays?: number
+  maxLockupTimeInDays?: number
+  maxLockupAmount: number
+  calcMultiplierFn: (lockupPeriodInDays: number) => number
+  onClose: () => void
+  onSubmit: (values: LockTokensModalFormValues) => Promise<void>
+}> = ({
+  isOpen = false,
+  mode = 'lock',
+  minLockupTimeInDays = 0,
+  maxLockupTimeInDays = Infinity,
+  maxLockupAmount,
+  calcMultiplierFn,
+  onClose,
+  onSubmit,
+}) => {
+  const { mint } = useRealm()
+  const [isSubmitting, setIsSubmitting] = useState(false)
+  const [showCustomDuration, setShowCustomDuration] = useState(false)
+  const [showLockupKindInfo, setShowLockupKindInfo] = useState<boolean>(false)
+  const mintMinAmount = mint ? getMintMinAmountAsDecimal(mint) : 1
+  const currentPrecision = precision(mintMinAmount)
+  const hasMinLockup = minLockupTimeInDays && minLockupTimeInDays > 0
+  const hasMaxLockup = maxLockupTimeInDays && maxLockupTimeInDays !== Infinity
+  const lockupKindOptions = [
+    { value: LockupKind.cliff, display: 'Cliff' },
+    { value: LockupKind.constant, display: 'Constant' },
+  ]
+
+  const lockupPeriodOptions = [
+    ...(hasMinLockup
+      ? [
+          {
+            value: minLockupTimeInDays,
+            display: 'min',
+          },
+        ]
+      : []),
+    ...defaultLockupPeriods.filter(
+      (lp) => lp.value > minLockupTimeInDays && lp.value <= maxLockupTimeInDays
+    ),
+  ]
+
+  const { setValue, watch, handleSubmit } = useForm<LockTokensModalFormValues>({
+    defaultValues: {
+      lockupKind: lockupKindOptions[0],
+      lockupPeriod: lockupPeriodOptions[0],
+      lockupPeriodInDays: lockupPeriodOptions[0].value,
+    },
+  })
+
+  const { lockupKind, amount, lockupPeriod, lockupPeriodInDays } = watch()
+
+  // validateAmountOnChange
+  useEffect(() => {
```

Comment:
(non blocking) Rather than validating using effects, it would probably be best to use react-hook-form's API. I found this article https://jasonwatmore.com/post/2021/04/21/react-hook-form-7-form-validation-example 


The reason I say this is because validating (and modifying) stuff while the user is in the middle of editing it often results in annoying jank.

Reply:
Was actually doing some onchange side effects in this logic. Just moved to the actual on change of the field.